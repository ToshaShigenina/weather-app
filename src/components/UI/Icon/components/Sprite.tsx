import React, { FC, SVGAttributes, useEffect } from 'react';
import { createPortal } from 'react-dom';
import { ReactComponent as SpriteSvg } from '../svg/sprite.svg'

import '../css/sprite.css';

const portalNode = document.createElement('div');
portalNode.setAttribute('id', 'sprite');
portalNode.setAttribute('data-testid', 'sprite');
portalNode.hidden = true;

export const Sprite: FC<SVGAttributes<SVGElement>> = (props) => {
	useEffect(() => {
		document.body.appendChild(portalNode);
		return () => portalNode.remove();
	});
	return createPortal(<SpriteSvg { ...props } />, portalNode)
};
