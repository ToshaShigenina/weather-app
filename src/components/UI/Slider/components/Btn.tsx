import React, { ButtonHTMLAttributes, FC } from 'react';
import cn from 'classnames';
import { ReactComponent as SliderArrow } from '../svg/slider-arrow.svg'

type SliderBtnDirectionType = 'prev' | 'next';
interface SliderBtnProps extends ButtonHTMLAttributes<HTMLButtonElement> {
    direction: SliderBtnDirectionType
}

export const Btn: FC<SliderBtnProps> = ({ direction = '', className, ...otherProps }) => {
    return (
        <button
            className={ cn(className, 'slider__btn', direction && `slider__btn-${direction}`) }
            { ...otherProps }
        >
            <SliderArrow />
        </button>
    );
};
