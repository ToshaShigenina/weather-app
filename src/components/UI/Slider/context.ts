import { createContext } from "react";

export type SliderContextType = {
	width: number
	margin: number
	sliderKeys: symbol[]
	addSliderKey?: (key: symbol) => void
}

export const SliderContext = createContext<SliderContextType | null>(null);
export const SliderContextProvider = SliderContext.Provider;
