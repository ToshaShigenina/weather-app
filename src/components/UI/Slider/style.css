.slider {
    position: relative;
    width: 100%;

}

.slider__wrapper {
    overflow: hidden;
}

.slider__container {
    display: flex;
    flex-wrap: nowrap;
    justify-content: flex-start;
    align-content: flex-start;
    transition: transform 0.5s;
}

.slider__btn {
    --color-bg: var(--color-white);
    --size: 32px;
    display: flex;
    align-items: center;
    justify-content: center;
    position: absolute;
    width: var(--size);
    height: var(--size);
    border-radius: 50%;
    background-color: var(--color-bg);
    border: none;
    margin-top: 15px;
    transform: translateY(-50%);
    cursor: pointer;
}

.slider__btn:disabled {
    opacity: 0.3;
    cursor: default;
    user-select: none;
}

.slider__btn-prev,
.slider__btn-next {
    top: calc(50% - (var(--size) / 2));
}

.slider__btn-prev {
    left: 0;
}

.slider__btn-next {
    right: 0;
}

.slider__btn-prev svg {
    transform: rotateZ(180deg);
}

.slide {
    flex-grow: 0;
    flex-shrink: 0;
    user-select: none;
}