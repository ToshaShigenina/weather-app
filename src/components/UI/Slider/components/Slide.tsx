import React, { FC, HTMLAttributes, useContext, useEffect, useState } from 'react';
import cn from 'classnames';
import { SliderContext, type SliderContextType } from '../context';

export const Slide: FC<HTMLAttributes<HTMLDivElement>> = ({ children, className = '', style = {}, ...otherProps }) => {
    const slideKey = Symbol();
    const [slideIndex, setSlideIndex] = useState(0);
    const sliderCtx = useContext<SliderContextType | null>(SliderContext);
    const width = sliderCtx && sliderCtx.width > 0
        ? sliderCtx.width + 'px'
        : '100%';
    const marginLeft = sliderCtx?.margin || 0;

    useEffect(() => {
        if (sliderCtx && sliderCtx.addSliderKey && !sliderCtx.sliderKeys.includes(slideKey)) {
            setSlideIndex(sliderCtx.sliderKeys.length);
            sliderCtx.addSliderKey(slideKey);
        }
    }, []);
    return (
        <div
            className={ cn(className, 'slide') }
            style={ {
                ...style,
                width,
                marginLeft: slideIndex ? marginLeft : 0,
            } }
            { ...otherProps }
        >
            { children }
        </div>
    );
};
