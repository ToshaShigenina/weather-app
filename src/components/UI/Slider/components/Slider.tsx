import React, { FC, HTMLAttributes, useRef, useState, useEffect } from 'react';
import { Btn } from './Btn';
import { useResize } from '../../../../hooks';
import { getCountSlides, getMinBreackpoint, getBreackpointsMap } from '../utils'
import { SliderContextProvider, type SliderContextType } from '../context';
import '../style.css';

export type SliderBreackpointProps = {
    slidesPerView?: number,
    marginBetweenSlides?: number,
}
export interface SliderProps extends HTMLAttributes<HTMLDivElement>, SliderBreackpointProps {
    breackpoints?: Record<number, SliderBreackpointProps>,
}

export const Slider: FC<SliderProps> = ({ children = null, slidesPerView = 1, marginBetweenSlides = 0, breackpoints, ...otherProps }) => {
    const sliderWrapperRef = useRef(null);
    const { width: windowWidth } = useResize();
    const [currentSlide, setCurrentSlide] = useState(1);
    const [slideWidth, setSlideWidth] = useState(0);
    const [countSlides, setCountSlides] = useState(0);
    const [countSlidesPerView, setCountSlidesPerView] = useState(slidesPerView);
    const [sliderGap, setSliderGap] = useState(marginBetweenSlides);
    const [translate, setTranslate] = useState(0);
    const [sliderKeys, setSliderKeys] = useState<symbol[]>([]);
    const context: SliderContextType = {
        width: slideWidth,
        margin: sliderGap,
        sliderKeys,
        addSliderKey: (key: symbol) => {
            sliderKeys.push(key);
            setSliderKeys(sliderKeys);
        }
    };

    const calcSlideWidth = () => {
        const count = getCountSlides(children);
        setCountSlides(count);
        if (count > 0 && sliderWrapperRef && sliderWrapperRef.current) {
            const sliderWidth = (sliderWrapperRef.current as HTMLDivElement).clientWidth
            setSlideWidth((sliderWidth - ((countSlidesPerView - 1) * sliderGap)) / countSlidesPerView);
        }
    };
    useEffect(() => {
        if (breackpoints) {
            const bpMap = getBreackpointsMap(breackpoints, windowWidth);
            if (Object.values(bpMap).every((item) => !item)) {
                setCountSlidesPerView(slidesPerView);
                setSliderGap(marginBetweenSlides);
            } else {
                const bpMin = getMinBreackpoint(bpMap);
                const bpTemp = breackpoints[bpMin];
                if (bpTemp.slidesPerView !== undefined) setCountSlidesPerView(bpTemp.slidesPerView);
                if (bpTemp.marginBetweenSlides !== undefined) setSliderGap(bpTemp.marginBetweenSlides);
            }
        }
        calcSlideWidth();
    }, [windowWidth]);
    useEffect(() => {
        calcSlideWidth();
    }, [children, countSlidesPerView, sliderGap]);

    const clickToPrevBtn = () => {
        if (currentSlide > 1) {
            setCurrentSlide(currentSlide - 1)
            setTranslate(translate + (slideWidth + sliderGap));
        }
    };
    const clickToNextBtn = () => {
        if (currentSlide <= countSlides - countSlidesPerView) {
            setCurrentSlide(currentSlide + 1)
            setTranslate(translate - (slideWidth + sliderGap));
        }
    };

    return (
        <SliderContextProvider value={ context }>
            <div className="slider" { ...otherProps }>
                <div className="slider__wrapper" ref={ sliderWrapperRef }>
                    <div
                        className="slider__container"
                        style={ { transform: `translateX(${translate}px)` } }
                    >
                        { children }
                    </div>
                </div>
                <Btn
                    onClick={ clickToPrevBtn }
                    direction="prev"
                    disabled={ currentSlide <= 1 }
                />
                <Btn
                    onClick={ clickToNextBtn }
                    direction="next"
                    disabled={ currentSlide > countSlides - countSlidesPerView }
                />
            </div>
        </SliderContextProvider>
    );
};
