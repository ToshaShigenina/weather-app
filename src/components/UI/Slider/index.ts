import { Slider as InitialSlider } from './components/Slider';
import { Slide } from './components/Slide';

type InitialSliderType = typeof InitialSlider;
type CompaundComponentType = InitialSliderType & {
	Slide: typeof Slide
};

const Slider = InitialSlider as CompaundComponentType;
Slider.Slide = Slide;

export { Slider };
export type { SliderBreackpointProps, SliderProps } from './components/Slider';