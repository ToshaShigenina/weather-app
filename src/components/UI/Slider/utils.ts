import { ReactNode } from 'react';
import { SliderBreackpointProps } from './components/Slider'

const asArray = (children: ReactNode): Array<ReactNode> => {
	if (children) {
		if (Array.isArray(children)) {
			return [...children];
		}
		return [children];
	}
	return [];
};
export const getCountSlides = (children: ReactNode): number => {
	return asArray(children).length
};
export const getMinBreackpoint = (breackpoints: Record<number, boolean>): number => {
	const bpTemp = Object.entries(breackpoints)
		.filter(([_, value]) => value)
		.map(([key, _]) => Number(key));
	return Math.min(...bpTemp);
};
export const getBreackpointsMap = (breackpoints: Record<number, SliderBreackpointProps>, width: number): Record<number, boolean> => {
	const keys = Object.keys(breackpoints);
	return keys.reduce((obj, key) => {
		if (!isNaN(Number(key))) {
			return {
				...obj,
				[key]: Number(key) > width
			}
		}
		return obj;
	}, {});
};