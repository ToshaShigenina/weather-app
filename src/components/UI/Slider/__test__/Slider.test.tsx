import React from 'react';
import { render, screen, fireEvent } from '@testing-library/react';
import { mountTest } from '../../../../tests';
import { Slider, type SliderProps } from '../components/Slider';
import { Slide } from '../components/Slide';

const testId = 'slider';
const testIdSlide = 'slide';

describe('Slider', () => {
	mountTest<SliderProps>(Slider);

	it('рендеринг пустого слайдера с дефолтными параметрами', () => {
		render(<Slider data-testid={ testId } />);
		const btns = screen.getAllByRole('button');
		expect(screen.getByTestId(testId)).toBeInTheDocument();
		expect(screen.getByTestId(testId)).toHaveClass('slider');
		expect(btns.length).toEqual(2);
		expect(btns[0]).toHaveClass('slider__btn', 'slider__btn-prev');
		expect(btns[1]).toHaveClass('slider__btn', 'slider__btn-next');
		expect(screen.getByTestId(testId)).toMatchSnapshot();
	});

	it('рендеринг слайдера с дефолтными параметрами', () => {
		const slidesContent = ['Slide 1', 'Slide 2', 'Slide 3'];
		render(
			<Slider data-testid={ testId }>
				{ slidesContent.map((text, i) => {
					return <Slide key={ i } data-testid={ testIdSlide }>{ text }</Slide>
				}) }
			</Slider>
		);
		const slides = screen.getAllByTestId(testIdSlide);

		expect(slides.length).toEqual(slidesContent.length);
		slides.forEach((item, i) => {
			expect(item).toHaveClass('slide');
			expect(item).toHaveStyle({
				width: '100%',
				'margin-left': '0px',
			});
			expect(item.textContent).toEqual(slidesContent[i]);
		});
		expect(screen.getByTestId(testId)).toMatchSnapshot();
	});

	it('рендеринг слайдера с отступами', () => {
		const slidesContent = ['Slide 1', 'Slide 2', 'Slide 3'];
		render(
			<Slider data-testid={ testId } marginBetweenSlides={ 10 }>
				{ slidesContent.map((text, i) => {
					return <Slide key={ i } data-testid={ testIdSlide }>{ text }</Slide>
				}) }
			</Slider>
		);
		const slides = screen.getAllByTestId(testIdSlide);

		slides.forEach((item, i) => {
			expect(item).toHaveStyle({
				width: '100%',
				'margin-left': !i ? '0px' : '10px',
			});
		});
		expect(screen.getByTestId(testId)).toMatchSnapshot();
	});

	it('слайды переключаются правильно', () => {
		const slidesContent = ['Slide 1', 'Slide 2'];
		render(
			<Slider data-testid={ testId }>
				{ slidesContent.map((text, i) => {
					return <Slide key={ i } data-testid={ testIdSlide }>{ text }</Slide>
				}) }
			</Slider>
		);
		const btns = screen.getAllByRole('button');

		expect(screen.getByTestId(testId)).toMatchSnapshot();
		expect(btns[0]).toBeDisabled();
		expect(btns[1]).not.toBeDisabled();
		fireEvent.click(btns[1]);
		expect(screen.getByTestId(testId)).toMatchSnapshot();
		expect(btns[0]).not.toBeDisabled();
		expect(btns[1]).toBeDisabled();
	});
});