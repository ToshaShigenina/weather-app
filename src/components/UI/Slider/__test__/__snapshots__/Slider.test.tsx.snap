// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Slider рендеринг пустого слайдера с дефолтными параметрами 1`] = `
<div
  class="slider"
  data-testid="slider"
>
  <div
    class="slider__wrapper"
  >
    <div
      class="slider__container"
      style="transform: translateX(0px);"
    />
  </div>
  <button
    class="slider__btn slider__btn-prev"
    disabled=""
  >
    <svg>
      slider-arrow.svg
    </svg>
  </button>
  <button
    class="slider__btn slider__btn-next"
    disabled=""
  >
    <svg>
      slider-arrow.svg
    </svg>
  </button>
</div>
`;

exports[`Slider рендеринг слайдера с дефолтными параметрами 1`] = `
<div
  class="slider"
  data-testid="slider"
>
  <div
    class="slider__wrapper"
  >
    <div
      class="slider__container"
      style="transform: translateX(0px);"
    >
      <div
        class="slide"
        data-testid="slide"
        style="width: 100%; margin-left: 0px;"
      >
        Slide 1
      </div>
      <div
        class="slide"
        data-testid="slide"
        style="width: 100%; margin-left: 0px;"
      >
        Slide 2
      </div>
      <div
        class="slide"
        data-testid="slide"
        style="width: 100%; margin-left: 0px;"
      >
        Slide 3
      </div>
    </div>
  </div>
  <button
    class="slider__btn slider__btn-prev"
    disabled=""
  >
    <svg>
      slider-arrow.svg
    </svg>
  </button>
  <button
    class="slider__btn slider__btn-next"
  >
    <svg>
      slider-arrow.svg
    </svg>
  </button>
</div>
`;

exports[`Slider рендеринг слайдера с отступами 1`] = `
<div
  class="slider"
  data-testid="slider"
>
  <div
    class="slider__wrapper"
  >
    <div
      class="slider__container"
      style="transform: translateX(0px);"
    >
      <div
        class="slide"
        data-testid="slide"
        style="width: 100%; margin-left: 0px;"
      >
        Slide 1
      </div>
      <div
        class="slide"
        data-testid="slide"
        style="width: 100%; margin-left: 10px;"
      >
        Slide 2
      </div>
      <div
        class="slide"
        data-testid="slide"
        style="width: 100%; margin-left: 10px;"
      >
        Slide 3
      </div>
    </div>
  </div>
  <button
    class="slider__btn slider__btn-prev"
    disabled=""
  >
    <svg>
      slider-arrow.svg
    </svg>
  </button>
  <button
    class="slider__btn slider__btn-next"
  >
    <svg>
      slider-arrow.svg
    </svg>
  </button>
</div>
`;

exports[`Slider слайды переключаются правильно 1`] = `
<div
  class="slider"
  data-testid="slider"
>
  <div
    class="slider__wrapper"
  >
    <div
      class="slider__container"
      style="transform: translateX(0px);"
    >
      <div
        class="slide"
        data-testid="slide"
        style="width: 100%; margin-left: 0px;"
      >
        Slide 1
      </div>
      <div
        class="slide"
        data-testid="slide"
        style="width: 100%; margin-left: 0px;"
      >
        Slide 2
      </div>
    </div>
  </div>
  <button
    class="slider__btn slider__btn-prev"
    disabled=""
  >
    <svg>
      slider-arrow.svg
    </svg>
  </button>
  <button
    class="slider__btn slider__btn-next"
  >
    <svg>
      slider-arrow.svg
    </svg>
  </button>
</div>
`;

exports[`Slider слайды переключаются правильно 2`] = `
<div
  class="slider"
  data-testid="slider"
>
  <div
    class="slider__wrapper"
  >
    <div
      class="slider__container"
      style="transform: translateX(0px);"
    >
      <div
        class="slide"
        data-testid="slide"
        style="width: 100%; margin-left: 0px;"
      >
        Slide 1
      </div>
      <div
        class="slide"
        data-testid="slide"
        style="width: 100%; margin-left: 0px;"
      >
        Slide 2
      </div>
    </div>
  </div>
  <button
    class="slider__btn slider__btn-prev"
  >
    <svg>
      slider-arrow.svg
    </svg>
  </button>
  <button
    class="slider__btn slider__btn-next"
    disabled=""
  >
    <svg>
      slider-arrow.svg
    </svg>
  </button>
</div>
`;
